docs/src/pages/components/data-grid/group-pivot/TreeDataLazyLoading.js

```diff
@@ -0,0 +1,359 @@
+        {row.descendantCount > 0 && (
+          <IconButton
+            size="small"
+            onClick={handleClick}
+            onKeyDown={handleKeyDown}
+            tabIndex={-1}
+            aria-label={
+              rowNode.childrenExpanded
+                ? apiRef.current.getLocaleText('treeDataCollapse')
+                : apiRef.current.getLocaleText('treeDataExpand')
+            }
+          >
+            <Icon fontSize="inherit" />
+          </IconButton>
+        )}
+      </div>
+      <span>
+        {rowNode.groupingKey}
+        {!hideDescendantCount && row.descendantCount > 0
+          ? ` (${row.descendantCount})`
+          : ''}
+      </span>
+    </Box>
+  );
+};
+
+GroupingCellWithLazyLoading.propTypes = {
+  /**
+   * The column field of the cell that triggered the event.
+   */
+  field: PropTypes.string.isRequired,
+  hideDescendantCount: PropTypes.bool,
+  /**
+   * The grid row id.
+   */
+  id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
+  /**
+   * The row model of the row that the current cell belongs to.
+   */
+  row: PropTypes.any.isRequired,
+  /**
+   * The node of the row that the current cell belongs to.
+   */
+  rowNode: PropTypes.shape({
+    /**
+     * The id of the row children.
+     */
+    children: PropTypes.arrayOf(
+      PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
+    ),
+    /**
+     * Current expansion status of the row.
+     */
+    childrenExpanded: PropTypes.bool,
+    /**
+     * 0-based depth of the row in the tree.
+     */
+    depth: PropTypes.number.isRequired,
+    /**
+     * The field used to group the children of this row.
+     * Is `null` if no field has been used to group the children of this row.
+     */
+    groupingField: PropTypes.oneOfType([PropTypes.oneOf([null]), PropTypes.string])
+      .isRequired,
+    /**
+     * The key used to group the children of this row.
+     */
+    groupingKey: PropTypes.oneOfType([
+      PropTypes.number,
+      PropTypes.string,
+      PropTypes.bool,
+    ]).isRequired,
+    /**
+     * The grid row id.
+     */
+    id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
+    /**
+     * If `true`, this node has been automatically added to fill a gap in the tree structure.
+     */
+    isAutoGenerated: PropTypes.bool,
+    /**
+     * The row id of the parent (null if this row is a top level row).
+     */
+    parent: PropTypes.oneOfType([
+      PropTypes.oneOf([null]),
+      PropTypes.number,
+      PropTypes.string,
+    ]).isRequired,
+  }).isRequired,
+};
+
+const CUSTOM_GROUPING_COL_DEF = {
+  renderCell: (params) => <GroupingCellWithLazyLoading {...params} />,
+};
+
+export default function TreeDataLazyLoading() {
+  const apiRef = useGridApiRef();
+  const [rows, setRows] = React.useState([]);
+
+  React.useEffect(() => {
+    fakeDataFetcher().then(setRows);
+
+    const handleRowExpansionChange = async (node) => {
```

Comment:
By curiosity, why putting it in a `useEffect` instead of a `useCallback` ?

Reply:
This `useEffect` is only called once to initialize the event listeners and init the rows.